streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,222 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
+        log.trace("Closing restoring stream tasks {}", revokedRestoringTasks);
+        RuntimeException exception = null;
+
+        final Iterator<TaskId> revokedTaskIterator = revokedRestoringTasks.iterator();
+        while (revokedTaskIterator.hasNext()) {
+            final TaskId id = revokedTaskIterator.next();
+            final StreamTask task = restoring.get(id);
+
+            if (task != null) {
+                log.debug("Closing restoring task {}", id);
+                try {
+                    task.closeStateManager(true);
+                } catch (final RuntimeException e) {
+                    log.error("Failed to close restoring task {} due to the following error:", id, e);
+                    if (exception == null) {
+                        exception = e;
+                    }
+                } finally {
+                    revokedTaskIterator.remove();
+                    restoring.remove(id);
+                    for (final TopicPartition tp : task.partitions()) {
+                        restoredPartitions.remove(tp);
+                        restoringByPartition.remove(tp);
+                    }
+                    revokedChangelogs.addAll(task.changelogPartitions());
+                }
+            } else {
+                // All non-restoring tasks should have been removed from revokedRestoringTasks, so log an error if it's not in restoring
+                log.error("Task was revoked but cannot be found in the assignment");
+                exception = new IllegalStateException("Revoked a task that was not in the assignment");
+                revokedTaskIterator.remove();
+            }
+        }
 
         return exception;
     }
 
+    /**
+     * @throws TaskMigratedException if the task producer got fenced (EOS only)
+     */
+    boolean maybeResumeSuspendedTask(final TaskId taskId, final Set<TopicPartition> partitions, final List<TopicPartition> changelogs) {
+        if (suspended.containsKey(taskId)) {
+            final StreamTask task = suspended.get(taskId);
+            log.trace("Found suspended {} {}", taskTypeName, taskId);
+            suspended.remove(taskId);
+
+            if (task.partitions().equals(partitions)) {
+                task.resume();
+                try {
+                    transitionToRunning(task);
+                } catch (final TaskMigratedException e) {
+                    // we need to catch migration exception internally since this function
+                    // is triggered in the rebalance callback
+                    log.info("Failed to resume {} {} since it got migrated to another thread already. " +
+                        "Closing it as zombie before triggering a new rebalance.", taskTypeName, task.id());
+                    final RuntimeException fatalException = closeZombieTask(task);
+                    running.remove(taskId);
+                    runningByPartition.keySet().removeAll(task.partitions());
```

Comment:
should we also remove all the changelog partitions? (We also add them in `transitionToRunning`).

Reply:
I'm not sure I understand what you're trying to say here and below, but: yes, we should technically add the changelogs here. Mostly to be safe since they will get cleared in onPartitionsLost since we have missed a rebalance. What do you mean we add them in transitionToRunning?

Reply:
Sorry, in the method `transitionToRunning`, I noticed that we add all the "partitions" _and_ all the "changelog partitions" to `runningByPartition`. But here, we only remove the "partitions". It seemed like symmetrically, we should also remove the "changelog partitions".

If the answer is that they get cleared anyway, then do we need to remove the "partitions" either?

Reply:
Ooh ok gotcha, yes we should remove the changelog partitions here and below. I do think it's best to keep the state consistent and not rely on an exact sequence of events.
That said, I'm thinking maybe we should not even try and close it as a zombie here and just let `onPartitionsLost` do its job?